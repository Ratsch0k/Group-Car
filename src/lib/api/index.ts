export * from './getRandomProfilePic';
export * from './group';
export * from './request.d';
export * from './userType.d';
export * from './auth';
export * from './user';
