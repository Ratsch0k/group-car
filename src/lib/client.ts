import axios, {AxiosResponse} from 'axios';

const csrfPromise = (() => {
  return axios.head('/auth').then((res: AxiosResponse) => {
    const csrf = res.headers['xsrf-token'];

    axios.defaults.headers.common = {
      'xsrf-token': csrf,
    };
  });
})();

const methodsToProxy = ['post', 'put', 'get', 'head', 'delete', 'options'];

methodsToProxy.forEach((method) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const origMethod = (axios as any)[method];

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (axios as any)[method] = (...args: any[]) => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return csrfPromise.then(() => origMethod(...args));
  };
});

export default axios;
