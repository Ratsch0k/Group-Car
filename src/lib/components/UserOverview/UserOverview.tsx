import {Button, Grid, Box, Typography} from '@material-ui/core';
import {createStyles, makeStyles} from '@material-ui/styles';
import AuthContext from 'lib/context/auth/authContext';
import React, {useContext, useState, useEffect} from 'react';
import UserAvatar from '../UserAvatar/UserAvatar';
import {useTranslation} from 'react-i18next';
import {grey} from '@material-ui/core/colors';

type GroupCarTheme = import('lib/theme').GroupCarTheme;

const useStyle = makeStyles((theme: GroupCarTheme) =>
  createStyles({
    container: {
      padding: theme.spacing(2),
    },
    userInfo: {
      border: `1px solid ${grey[500]}`,
      borderRadius: theme.shape.borderRadius,
      marginBottom: theme.spacing(4),
      padding: theme.spacing(1),
      minWidth: 180,
    },
  }),
);

interface UserOverviewProps {
  onClose?(): void;
}

const UserOverview: React.FC<UserOverviewProps> =
(props: UserOverviewProps) => {
  const auth = useContext(AuthContext);

  /**
   * Store user info in state so that it doesn't
   * change while the popover for this component closes.
   */
  const [userInfo, setUserInfo] = useState<JSX.Element | null>(null);

  const {t} = useTranslation();

  const classes = useStyle();

  const handleLogout = () => {
    props.onClose && props.onClose();
    auth.logout();
  };

  useEffect(() => {
    // Only update the user info if the user changes
    if (auth.isLoggedIn && auth.user) {
      setUserInfo(
          <>
            <UserAvatar userId={auth.user && auth.user.id} size={100}/>
            <Typography align='center'>
              {auth.user && auth.user.username}
            </Typography>
          </>,
      );
    }
  }, [auth.user]);

  return (
    <Box className={classes.container}>
      <Grid
        container
        direction='column'
        justify='flex-start'
        alignItems='center'
      >
        <Grid
          item
          className={classes.userInfo}
        >
          {userInfo}
        </Grid>
        <Grid item>
          <Button
            data-testid='logout-button'
            color='primary'
            variant='contained'
            onClick={handleLogout}
            fullWidth
          >
            {t('form.logout')}
          </Button>
        </Grid>
      </Grid>
    </Box>
  );
};

export default UserOverview;
