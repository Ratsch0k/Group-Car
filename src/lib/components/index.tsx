export * from './Authentication';
export * from './CloseableDialogTitle';
export * from './Drawer';
export * from './DrawerCard';
export * from './HeaderBar';
export * from './Input';
export * from './UserAvatar';
export * from './UserOverview';
export * from './legal';
export * from './AutoFullscreenDialog';
export * from './AxiosErrorHandler';
