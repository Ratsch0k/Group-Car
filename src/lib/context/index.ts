export * from './axiosContext';
export * from './apiContext';
export * from './invitesContext';
export * from './snackbarContext';
export * from './mapContext';
