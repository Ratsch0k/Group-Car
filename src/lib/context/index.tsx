export * from './authContext';
export * from './axiosContext';
