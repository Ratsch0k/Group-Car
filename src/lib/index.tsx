export * from './components';
export * from './ModalRouter';
export * from './theme';
export * from './requests';
export * from './Routes';
export * from './context';
export * from './hooks';
