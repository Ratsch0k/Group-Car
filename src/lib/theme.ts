import {createMuiTheme} from '@material-ui/core';

const theme = createMuiTheme({
  palette: {
    primary: {
      main: '#00C68A',
      light: '#98F0D6',
      dark: '#00A473',
      contrastText: '#FFFFFF',
    },
    secondary: {
      main: '#AD00D8',
      light: '#ECA1FF',
      dark: '#8900AB',
      contrastText: '#FFFFFF',
    },
  },
});

export default theme;
