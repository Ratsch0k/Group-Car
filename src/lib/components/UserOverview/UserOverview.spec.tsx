import React from 'react';
import {render, fireEvent, screen} from '@testing-library/react';
import UserOverview from './UserOverview';
import {ThemeProvider} from '@material-ui/core';
import {AuthContext} from 'lib';
import theme from '../../../__test__/testTheme';


it('renders and matches snapshot when user is not logged in', () => {
  const fakeContext = {
    login: jest.fn(),
    signUp: jest.fn(),
    logout: jest.fn(),
    user: undefined,
    isLoggedIn: false,
    openAuthDialog: jest.fn(),
  };

  const {container} = render(
      <ThemeProvider theme={theme}>
        <AuthContext.Provider value={fakeContext}>
          <UserOverview />
        </AuthContext.Provider>
      </ThemeProvider>,
  );

  expect(container).toMatchSnapshot();
  expect(fakeContext.login).not.toHaveBeenCalled();
  expect(fakeContext.signUp).not.toHaveBeenCalled();
  expect(fakeContext.logout).not.toHaveBeenCalled();
  expect(fakeContext.openAuthDialog).not.toHaveBeenCalled();
});

it('renders and matches snapshot when user is logged in', () => {
  const fakeUser = {
    id: 12,
    username: 'TEST',
    email: 'TEST@mail.com',
    isBetaUser: false,
  };

  const fakeContext = {
    login: jest.fn(),
    signUp: jest.fn(),
    logout: jest.fn(),
    user: fakeUser,
    isLoggedIn: true,
    openAuthDialog: jest.fn(),
  };

  const {container} = render(
      <ThemeProvider theme={theme}>
        <AuthContext.Provider value={fakeContext}>
          <UserOverview />
        </AuthContext.Provider>
      </ThemeProvider>,
  );

  expect(container).toMatchSnapshot();
  expect(fakeContext.login).not.toHaveBeenCalled();
  expect(fakeContext.signUp).not.toHaveBeenCalled();
  expect(fakeContext.logout).not.toHaveBeenCalled();
  expect(fakeContext.openAuthDialog).not.toHaveBeenCalled();
});

it('logout calls props.onClose and auth.logout', () => {
  const fakeUser = {
    id: 12,
    username: 'TEST',
    email: 'TEST@MAIL.COM',
    isBetaUser: false,
  };

  const fakeContext = {
    login: jest.fn(),
    signUp: jest.fn(),
    logout: jest.fn(),
    user: fakeUser,
    isLoggedIn: true,
    openAuthDialog: jest.fn(),
  };

  const onClose = jest.fn();

  render(
      <ThemeProvider theme={theme}>
        <AuthContext.Provider value={fakeContext}>
          <UserOverview onClose={onClose}/>
        </AuthContext.Provider>
      </ThemeProvider>,
  );

  fireEvent.click(screen.getByTestId('logout-button'));

  expect(fakeContext.login).not.toHaveBeenCalled();
  expect(fakeContext.signUp).not.toHaveBeenCalled();
  expect(fakeContext.logout).toHaveBeenCalledTimes(1);
  expect(fakeContext.openAuthDialog).not.toHaveBeenCalled();
  expect(onClose).toHaveBeenCalled();
});
