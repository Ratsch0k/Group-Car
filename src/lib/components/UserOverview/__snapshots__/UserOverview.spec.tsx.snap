// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders and matches snapshot when user is logged in 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-153 makeStyles-container-150"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column"
    >
      <div
        class="MuiGrid-root makeStyles-userInfo-151 MuiGrid-item"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-289"
        >
          <img
            alt="user-12"
            class="MuiAvatar-img"
            src="/api/user/12/profile-pic"
          />
        </div>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter"
        >
          TEST
        </p>
        <span
          class="MuiTypography-root MuiTypography-caption MuiTypography-colorTextSecondary MuiTypography-alignCenter"
        />
      </div>
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
          data-testid="logout-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            form.logout
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders and matches snapshot when user is not logged in 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-4 makeStyles-container-1"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column"
    >
      <div
        class="MuiGrid-root makeStyles-userInfo-2 MuiGrid-item"
      />
      <div
        class="MuiGrid-root MuiGrid-item"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
          data-testid="logout-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            form.logout
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
